### DISCUSSION TITLE
Handling of random numbers

### DISCUSSION THREAD

**Discussion by DougBurke** [sherpa/sherpa#1742] [Ideas]
At the moment we do not have a unified interface to random numbers:

- some areas take a `seed` argument (mainly in the optimizer code) which is normally accessible to the user in some way
- some routines require you to call `np.random.seed(seed)` to be repeatable
- some routines require you to call `random.seed(seed)` to be repeatable
- there's probably some routines that need both `np.random.seed` and `random.seed` to be set (I have not checked this, but I think it likely to be the case from some of the bizarre issues I see when writing tests and checking their coverage)
- we don't really document any of this
- it is not clear to me how the existing code works when using `multiprocessing` to run code that needs random numbers in parallel

With NumPy 1.17 now being accessible for all useful builds (as in, it's actually now out of date) we can move to an interface where we send around a generator rather than rely on hidden global state (which `np.random.seed` and `random.seed` do) - which is essentially issue #867.

This would

- provide a uniform interface
- automatically document those routines that use random numbers (although there's arguments to be bad for the `sherpa.ui/sherpa.astro.ui` routines that we should treat such a generator as we do the choice of statistic or optimizer)

So, my suggestion is that we add a `rng` parameter which can be one of

- `None` - this would currently rely on the existing "global" state used by the legacy `np.random` code and the `random` module, to allow existing code to work, but eventually would switch to meaning "just call `np.random.default_rng()`" 
- send in a NumPy generator - e.g. `np.random.default_rng()`
- send in a `np.random.RandomState` object, which is the equivalent of the "legacy" numpy random interface, and is primarily useful for testing, as it is guaranteed to produce the same sequence of numbers

There are a number of issues with this, including

- how does this interact with those routines which accept a `seed` argument?
  - my thinking is that if `rng is None` then we use the sent in `seed` value, otherwise we use the generator to create a new seed value to use in the routine instead of the user-supplied value. We can mark the `seed` argument as deprecated in some of these routines, but for some code it is required, since this seed is sent in to C++ scope where it is used to seed a [Mersenne-Twister](https://en.wikipedia.org/wiki/Mersenne_Twister) based implementation which we are not going to re-written to get to call back into Python scope to get random numbers
- we can clarify the random-number-access-in-parallel by using the `np.random.SeedSequence` class to create a separate generator for each process. I would like to be able to create a generator that matches the sent-in version so that if you've selected a particular generator rather than using `np.random.default_rng` we can continue using this, but it's not obvious that the `Generator` API allows this (I still think this is an improvement over the existing code, if only conceptually)

This is based on my investigations in #1735 (at least at the time of writing - April 7 2023). This is a large-enough change that it would make sense to have community input - e.g. how does AstroPy handle passing around the random generators?

**Comment by hamogu** [comment]
To the last point: Astropy rarely uses random numbers. I can't actually think of another similarly-sized package that uses random numbers "a bit" like Sherpa. Either it's all about random numbers (e.g. PyMC) or random numbers just turn up in a handful or routines (then you don't have to bother with a consistent interface)

**Comment by DougBurke** [comment]
With the 4.16.0 release we are now a bit better - low-level routines should take a `rng` argument and we no-longer rely on math.random, just the NumPy RNG (either the new style generators or the old-style single seed) and UI routines have `set_rng`/`get_rng` for setting / gettnig the RNG argument to use.

There is still the use of `seed` for the optimisation code (as part of the settings).
